### ISSUE TITLE
`@typescript-eslint/promise-function-async` should not flag Promise subclasses

### ISSUE THREAD

**Issue by chancancode** [typescript-eslint/typescript-eslint#2034]
Functions that return promises must be async.eslint@typescript-eslint/promise-function-async
The return type of an async function or method must be the global Promise<T> type.ts(1064)

**Repro**

[Playground](https://typescript-eslint.io/play#ts=5.3.2&fileType=.tsx&code=MYGwhgzhAECyCeAFATgewLYEsIFNo4A8AXHAOwBMYUNscAeAN1U3ID5oBvAKGmgHMcRaNgDKABxzBMYEAAoAlAC5oAI1SoQOMKU49e0ZIICuyHUWRGcAbj0BfLva4AzI6WBFMqHU-ULlCaixcXV5DIhMdUhwAdzgkNCCcWQVoAF52DmhbeRt7IA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6AewFtLlFaAM1hMy%2BSpya0AhskKj0YANrhsORNC7RIqiAF1VAXxCGgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

```JSON
{
  "rules": {
    "@typescript-eslint/promise-function-async": ["error"],
  }
}
```

```TS
class MyPromise extends Promise<void> {
  get isSpecial(): boolean {
    return true;
  }
}

function foo(): MyPromise {
  return new MyPromise(() => {});
}
```

**Expected Result**

No error.

**Actual Result**

```
Functions that return promises must be async.eslint@typescript-eslint/promise-function-async
```

**Additional Info**

I understand that there is an `allowedPromiseNames` configuration. However, IMO, this is an invalid error and an invalid recommendation.

By spec, async function always and can only return exactly the built-in `Promise` objects. `async function foo(): MyPromise { ... }` is not valid code and if you try to write that, TypeScript will yell at you with `The return type of an async function or method must be the global Promise<T> type.ts(1064)`.

If you wrote `function foo(): MyPromise`, you are specifically saying you are not returning just a `Promise<T>`, therefore it is not an async function. Therefore, I believe this is a false positive picked up by the rule. It should not warn about this case. IMO, the `allowedPromiseNames` config should also be removed, I don't think there is a valid use case for it, and if you do that you will get a type error from TS anyway.

**Versions**

| package                                   | version |
| ----------------------------------------- | ------- |
| `@typescript-eslint/parser`               | `2.31.0` |
| `TypeScript`                              | `3.8.3` |
| `ESLint`                                  | `5.16.0` |
| `node`                                    | `12.16.3` |
| `yarn`                                     | `1.22.4` |

**Comment by bradzacher** [comment]
It's not entirely correct.
If you're targeting ES5 with your TS code, then you can use whatever Promise implementation that you like.

It's only if you're targeting a higher ES version that you can't.

The rule has been around for a long time (#194), and even then it was a direct port of the TSLint rule which has been around even longer (https://github.com/palantir/tslint/pull/1779).
Checking out the source, the rule hasn't really changed since its first implementation in 2016. It's not been a problem because 99% of people don't do custom promises.

It probably could do with a check now to make it understand that a target higher than ES5 means you can only use the global `Promise` type defined by TS.

As mentioned - this isn't a problem for almost all our users, so happy to accept a PR if you want to see this functionality.

**Comment by KleaTech** [comment]
The `allowedPromiseNames` cannot be used to work around this issue. That option has a misleading name, it does not turn off the rule for the given names, it will turn on the rule for them. So if you add a type to `allowedPromiseNames` that is not a subclass of Promise, the rule will act as it would be a subclass of Promise. If you add a type to `allowedPromiseNames` that is already a subclass of Promise, it will do nothing, the rule would still trigger.

Therefor there's no workaround for this issue (except disabling the rule completely).

**Comment by JoshuaKGoldberg** [comment]
This has been accepting PRs for over 2 years and has only received positive interactions from two community members. That indicates it's not likely worth the breaking change or maintenance cost for us long-term. Closing out as wontfix.

If anybody really wants this rule, I'd encourage using the custom rules tooling documented at https://typescript-eslint.io/developers/custom-rules to build it and make it available as a standalone plugin.

Thanks all!
